libs/pika/async_mpi/tests/unit/mpi_ring_async_sender_receiver.cpp

```diff
@@ -40,169 +58,425 @@ namespace mpi = pika::mpi::experimental;
 namespace tt = pika::this_thread::experimental;
 namespace deb = pika::debug;
 
-static bool output = true;
+static bool output = false;
+static uint32_t mpi_task_transfer = mpi::get_completion_mode();
+static std::uint32_t mpi_poll_size = 16;
+std::atomic<std::uint64_t> counter;
+
+// ------------------------------------------------------------
+// a debug level of zero disables messages with a priority>0
+// a debug level of N shows messages with priority<N
+using namespace pika::debug::detail;
+template <int Level>
+static print_threshold<Level, 1> msr_deb("MPI_SR_");
+
+// ------------------------------------------------------------
+// caution: message_buffers will be constructed in-place in a buffer
+// allocated with a larger size (set by message-bytes option)
+struct message_buffer
+{
+    std::uint64_t token_val_;
+    std::uint64_t size_;
+    //
+    std::uint64_t size()
+    {
+        return size_;
+    }
+    //
+    message_buffer(std::uint64_t msize)
+    {
+        token_val_ = 65535;
+        size_ = msize;
+    }
+};
+
+// ------------------------------------------------------------
+inline std::uint32_t next_rank(std::uint32_t rank, std::uint32_t size)
+{
+    return (rank + 1) % size;
+}
+
+inline std::uint32_t prev_rank(std::uint32_t rank, std::uint32_t size)
+{
+    return (rank + size - 1) % size;
+}
 
-void msg_recv(int rank, int size, int /*to*/, int from, int token, unsigned tag)
+// ------------------------------------------------------------
+// when messages arrive/complete in random order, it might be the
+// case that message N + x completes long before message N, so
+// we don't reuse tags immediately and instead offset them by
+// a multiple to ensure we don't have two with the same id in flight
+// at the same time
+inline std::uint32_t tag_no(std::uint64_t rank, std::uint64_t iteration, std::uint32_t ranks)
```

Comment:
Maybe?
```suggestion
inline std::uint32_t tag(_num(ber))(std::uint64_t rank, std::uint64_t iteration, std::uint32_t ranks)
```
`tag_no` sounds pretty negative to me... :wink: